dom.go

```diff
@@ -297,61 +320,120 @@ func Text(text string, m ...MarkupOrComponentOrHTML) *HTML {
 }
 
 // Rerender causes the body of the given component (i.e. the HTML returned by
-// the Component's Render method) to be re-rendered and subsequently restored.
+// the Component's Render method) to be re-rendered.
+//
+// If the component has not been rendered before, Rerender panics.
 func Rerender(c Component) {
+	if c == nil {
+		panic("vecty: Rerender illegally called with a nil Component argument")
+	}
 	prevRender := c.Context().prevRender
-	nextRender := doRender(c)
-	var prevComponent Component // TODO
-	if doRestore(prevComponent, c, prevRender, nextRender) {
-		return
+	if prevRender == nil {
+		panic("vecty: Rerender invoked on Component that has never been rendered")
 	}
-	if prevRender != nil {
-		replaceNode(nextRender.node, prevRender.node)
+	nextRender, skip := doRender(c)
+	if skip {
+		return
 	}
+	replaceNode(nextRender.node, prevRender.node)
 }
 
-func doRender(c ComponentOrHTML) *HTML {
-	if h, isHTML := c.(*HTML); isHTML {
-		return h
+// doCopy makes a copy of the given component.
+func doCopy(c Component) Component {
+	if c == nil {
+		panic("vecty: cannot copy nil Component")
 	}
-	comp := c.(Component)
-	r := renderHandleNil(comp)
-	comp.Context().prevRender = r
-	return r
+
+	// If the Component implements the Copier interface, then use that to
+	// perform the copy.
+	if copier, ok := c.(Copier); ok {
+		cpy := copier.Copy()
+		if cpy == c {
+			panic("vecty: Component.Copy returned an identical *MyComponent pointer")
+		}
+		return cpy
+	}
+
+	// Component does not implement the Copier interface, so perform a shallow
+	// copy.
+	v := reflect.ValueOf(c)
+	if v.Kind() != reflect.Ptr || v.Elem().Kind() != reflect.Struct {
+		panic(fmt.Sprintf("vecty: Component must be pointer to struct, found %T", c))
+	}
+	cpy := reflect.New(v.Elem().Type())
+	cpy.Elem().Set(v.Elem())
+	return cpy.Interface().(Component)
 }
 
-func renderHandleNil(c Component) *HTML {
-	r := c.Render()
-	if r == nil {
-		// nil renders are translated into noscript tags.
-		r = Tag("noscript")
+// doRenderHTML renders the ComponentOrHTML. In the case of *HTML, its Restore
+// method is invoked with the specified prevRender and c.(*HTML) is returned.
+//
+// In the case of a Component, doRender(c.(Component)) is returned.
+func doRenderHTML(c ComponentOrHTML, prevRender *HTML) (h *HTML, skip bool) {
+	switch v := c.(type) {
+	case *HTML:
+		v.Restore(prevRender)
+		return v, false
+	case Component:
+		return doRender(v)
+	default:
+		panic(fmt.Sprintf("vecty: encountered invalid ComponentOrHTML %T", c))
 	}
-	return r
 }
 
-func doRestore(prev, next ComponentOrHTML, prevRender, nextRender *HTML) (skip bool) {
-	if r, ok := next.(Restorer); ok {
-		var p Component
+// doRender handles rendering the given Component into *HTML. If skip == true
+// is returned, the Component's SkipRender method has signaled the component does
+// not need to be rendered and h == nil is returned.
+func doRender(comp Component) (h *HTML, skip bool) {
```

Comment:
Good catch. Not sure why i didn't think of that when working on this. I agree it would be better to move the component `Restore` call above the `SkipRender` call here. Should also add a test to make sure that it is invoked in that order to prevent any future change from altering the order without considering that case.